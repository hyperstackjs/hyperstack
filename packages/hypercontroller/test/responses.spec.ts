import request from 'supertest'

import type { Request, Response } from '../src/index'
import {
  Controller,
  Get,
  HttpResponseOK,
  ValidationError,
  createServer,
} from '../src/index'
import { expectWithSnapshot, logger } from './utils'
@Controller('api')
class Api {
  @Get('foobar')
  async foobar(_req: Request, _res: Response) {
    return new HttpResponseOK({ its: 'alright' })
  }

  @Get('no-reqres')
  async noreqres() {
    return new HttpResponseOK({ its: 'alright' })
  }

  @Get('throw-me')
  async throwme(_req: Request, _res: Response) {
    throw new Error('it cant work')
  }

  @Get('validation-error')
  async validationErrors(_req: Request, _res: Response) {
    throw new ValidationError({ errors: [] })
  }

  @Get('uri-error')
  async uriErrors(_req: Request, _res: Response) {
    throw new URIError("this is a uri error")
  }

  @Get('raw-object')
  async rawObject(_req: Request, _res: Response) {
    return { hello: 'ok' }
  }
}

const app = createServer({
  controllers: [Api],
  opts: { logging: { logger }, sendValidationErrors: true },
})
describe('hypercontroller/responses', () => {
  it('requests', async () => {
    await expectWithSnapshot(200, request(app).get('/api/foobar'))
  })
  it('no req res', async () => {
    await expectWithSnapshot(200, request(app).get('/api/no-reqres'))
  })
  it('wraps async erors', async () => {
    await expectWithSnapshot(400, request(app).get('/api/throw-me'))
  })
  it('validation error', async () => {
    await expectWithSnapshot(400, request(app).get('/api/validation-error'))
  })
  it('uri error', async () => {
    await expectWithSnapshot(400, request(app).get('/api/uri-error'))
  })
  it('raw object', async () => {
    await expectWithSnapshot(200, request(app).get('/api/raw-object'))
  })
})
