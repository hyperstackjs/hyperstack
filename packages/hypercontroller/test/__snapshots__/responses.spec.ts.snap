// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hypercontroller/responses no req res 1`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`hypercontroller/responses raw object 1`] = `
{
  "body": {
    "hello": "ok",
  },
  "headers": {
    "connection": "close",
    "content-length": "14",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`hypercontroller/responses requests 1`] = `
{
  "body": {
    "its": "alright",
  },
  "headers": {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`hypercontroller/responses uri error 1`] = `
{
  "body": {
    "error": "this is a uri error",
  },
  "headers": {
    "connection": "close",
    "content-length": "31",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`hypercontroller/responses validation error 1`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": {
      "errors": [],
    },
  },
  "headers": {
    "connection": "close",
    "content-length": "55",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`hypercontroller/responses wraps async erors 1`] = `
{
  "body": {
    "error": "bad request",
  },
  "headers": {
    "connection": "close",
    "content-length": "23",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;
